File: src/lib/types/session.ts
export type SessionState =
	| 'running'
	| 'needs-input'
	| 'needs-review'
	| 'paused'
	| 'finished'
	| 'errored';

export type ExecutionPhase =
	| 'none'
	| 'analyzing'
	| 'tdd'
	| 'reviewing'
	| 'verifying'
	| 'committing';

export interface Session {
	id: string;
	issue_id: string | null;
	provider: string;
	state: SessionState;
	pid: number | null;
	session_file_path: string | null;
	started_at: string;
	ended_at: string | null;
	cost_usd: number | null;
	token_count: number | null;
	original_intent: string | null;
	last_prompt: string | null;
	last_response_summary: string | null;
	execution_phase: ExecutionPhase;
	source: 'spawned' | 'adopted';
	working_directory: string | null;
}

export interface SpawnSessionRequest {
	prompt: string;
	working_directory: string;
	issue_id?: string | null;
	permission_mode?: string | null;
	model?: string | null;
}

export type SessionEvent =
	| { type: 'session_init'; session_id: string; model: string; tools: string[] }
	| { type: 'message_delta'; text: string }
	| { type: 'message_complete'; text: string; message_id: string }
	| { type: 'thinking_delta'; text: string }
	| { type: 'tool_start'; tool_use_id: string; tool_name: string; input: unknown }
	| {
			type: 'tool_end';
			tool_use_id: string;
			tool_name: string;
			output: unknown;
			is_error: boolean;
	  }
	| { type: 'tool_progress'; tool_use_id: string; elapsed_seconds: number }
	| { type: 'tool_use_summary'; tool_use_id: string; summary: string }
	| { type: 'run_state'; state: string; error: string | null }
	| { type: 'usage_update'; input_tokens: number; output_tokens: number; cost_usd: number }
	| { type: 'permission_prompt'; request_id: string; tool_name: string; tool_input: unknown }
	| { type: 'elicitation_prompt'; request_id: string; message: string }
	| { type: 'compact_boundary'; trigger: string }
	| { type: 'system_status'; status: string }
	| { type: 'control_cancelled'; request_id: string }
	| { type: 'raw'; source: string; data: unknown };

export type SessionEventType = SessionEvent['type'];

export interface SessionEventPayload {
	session_id: string;
	event: SessionEvent;
}

// ─── Discovered Session Types ──────────────────────────────────────────────

export type DiscoveredSessionStatus =
	| 'working'
	| 'needs_attention'
	| 'idle'
	| 'finished'
	| 'unknown';

export interface DiscoveredSession {
	id: string;
	pid: number;
	working_directory: string;
	project_directory_name: string;
	session_id: string;
	project_name: string;
	status: DiscoveredSessionStatus;
	first_prompt: string | null;
	git_branch: string | null;
	message_count: number;
	cost_usd: number;
	token_count: number;
	latest_message: string | null;
	modified_at: string | null;
}

export interface DiscoveredSessionsPayload {
	sessions: DiscoveredSession[];
}

export interface AdoptSessionRequest {
	cli_session_id: string;
	working_directory: string;
	issue_id?: string | null;
	original_intent?: string | null;
	cost_usd?: number | null;
	token_count?: number | null;
}

File: src-tauri/src/models/session.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS)]
#[ts(export)]
#[serde(rename_all = "kebab-case")]
pub enum SessionState {
    Running,
    NeedsInput,
    NeedsReview,
    Paused,
    Finished,
    Errored,
}

impl_sql_enum!(SessionState {
    Running => "running",
    NeedsInput => "needs-input",
    NeedsReview => "needs-review",
    Paused => "paused",
    Finished => "finished",
    Errored => "errored",
});

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS)]
#[ts(export)]
#[serde(rename_all = "kebab-case")]
pub enum ExecutionPhase {
    None,
    Analyzing,
    Tdd,
    Reviewing,
    Verifying,
    Committing,
}

impl_sql_enum!(ExecutionPhase {
    None => "none",
    Analyzing => "analyzing",
    Tdd => "tdd",
    Reviewing => "reviewing",
    Verifying => "verifying",
    Committing => "committing",
});

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS)]
#[ts(export)]
#[serde(rename_all = "kebab-case")]
pub enum SessionSource {
    Spawned,
    Adopted,
}

impl_sql_enum!(SessionSource {
    Spawned => "spawned",
    Adopted => "adopted",
});

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct Session {
    pub id: String,
    pub issue_id: Option<String>,
    pub provider: String,
    pub state: SessionState,
    #[ts(type = "number | null")]
    pub pid: Option<i64>,
    pub session_file_path: Option<String>,
    pub started_at: String,
    pub ended_at: Option<String>,
    pub cost_usd: Option<f64>,
    #[ts(type = "number | null")]
    pub token_count: Option<i64>,
    pub original_intent: Option<String>,
    pub last_prompt: Option<String>,
    pub last_response_summary: Option<String>,
    pub execution_phase: ExecutionPhase,
    pub source: SessionSource,
    pub working_directory: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SpawnSessionRequest {
    pub prompt: String,
    pub working_directory: String,
    pub issue_id: Option<String>,
    pub permission_mode: Option<String>,
    pub model: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn session_state_serde_round_trip() {
        let json = "\"needs-input\"";
        let deserialized: SessionState = serde_json::from_str(json).unwrap();
        assert_eq!(deserialized, SessionState::NeedsInput);
        let serialized = serde_json::to_string(&deserialized).unwrap();
        assert_eq!(serialized, "\"needs-input\"");
    }

    #[test]
    fn execution_phase_serde_round_trip() {
        let json = "\"tdd\"";
        let deserialized: ExecutionPhase = serde_json::from_str(json).unwrap();
        assert_eq!(deserialized, ExecutionPhase::Tdd);
        let serialized = serde_json::to_string(&deserialized).unwrap();
        assert_eq!(serialized, "\"tdd\"");
    }

    #[test]
    fn session_source_serde_round_trip() {
        let json = "\"spawned\"";
        let deserialized: SessionSource = serde_json::from_str(json).unwrap();
        assert_eq!(deserialized, SessionSource::Spawned);
        let serialized = serde_json::to_string(&deserialized).unwrap();
        assert_eq!(serialized, "\"spawned\"");
    }

    #[test]
    fn invalid_session_state_deserialization_returns_error() {
        let json = "\"invalid-state\"";
        let result = serde_json::from_str::<SessionState>(json);
        assert!(result.is_err());
    }
}
